frontends/ast/simplify.cc

```diff
@@ -3089,6 +3089,138 @@ skip_dynamic_range_lvalue_expansion:;
 				goto apply_newNode;
 			}
 
+			if (str == "\\$countbits") {
+				if (children.size() < 2)
+					log_file_error(filename, location.first_line, "System function %s got %d arguments, expected at least 2.\n",
+							RTLIL::unescape_id(str).c_str(), int(children.size()));
+
+				bool count_0 = false;
+				bool count_1 = false;
+				bool count_x = false;
+				bool count_z = false;
+
+				/* Determine which bits to count */
+				for (size_t i = 1; i < children.size(); i++) {
+					AstNode *node = children[i];
+					while (node->simplify(true, false, false, stage, -1, false, false)) { }
+					if (node->type != AST_CONSTANT)
+						log_file_error(filename, location.first_line, "Failed to evaluate system function `%s' with non-constant control bit argument.\n", str.c_str());
+					if (node->bits.size() != 1)
+						log_file_error(filename, location.first_line, "Failed to evaluate system function `%s' with control bit width != 1.\n", str.c_str());
+					switch (node->bits[0]) {
+						case RTLIL::State::S0: count_0 = true; break;
+						case RTLIL::State::S1: count_1 = true; break;
+						case RTLIL::Sx:        count_x = true; break;
+						case RTLIL::Sz:        count_z = true; break;
+						default:;
+					}
+				}
+
+				/* Detect width of exp (first argument of $countbits) */
+				int  exp_width = -1;
+				bool exp_sign  = false;
+				AstNode *exp = children[0];
+				exp->detectSignWidth(exp_width, exp_sign, NULL);
+
+				for (int i = 0; i < exp_width; i++) {
+					std::vector<RTLIL::State> mask(exp_width, RTLIL::State::S0);
+
+					/* Generate node for:  exp & ('1 << i) */
+					/*                         ^           */
+					mask[i] = RTLIL::State::S1;
+					AstNode *and_node = new AstNode(AST_BIT_AND, exp->clone(), mkconst_bits(mask, false));
+
+					AstNode *eq0_node;
+					AstNode *eq1_node;
+					AstNode *eqx_node;
+					AstNode *eqz_node;
+
+					if (count_0) {
+						/* Generate node for:  (exp & ('1 << i)) === ('0 << i) */
+						/*                                       ^^^           */
+						mask[i] = RTLIL::State::S0;
+						eq0_node = new AstNode(AST_EQX, and_node->clone(), mkconst_bits(mask, false));
+					} else {
+						eq0_node = mkconst_int(0, false);
+					}
```

Comment:
I think these cases may benefit from abstraction into a lambda.

Reply:
I collected the bit states in a vector so I can iterate over it instead of the four count_x vars.

The shifting works for 'z. Thanks.